File: src/main/java/com/nekoloop/base64image/Base64Image.java
package com.nekoloop.base64image;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

/**
 * Created by Nguyen Hoang Anh Nguyen on 2/16/16.
 */
public class Base64Image {

    static final Handler HANDLER = new Handler(Looper.getMainLooper()) {
        @Override
        public void handleMessage(Message msg) {
            switch (msg.what) {
                default:
                    throw new AssertionError("Unknown handler message received: " + msg.what);
            }
        }
    };

    private static volatile Base64Image singleton = null;

    final Context context;
    final ExecutorService service;
    final RequestDecode.Provider base64Provider;
    final RequestEncode.Provider bitmapProvider;

    private Base64Image(Context context, ExecutorService service, RequestDecode.Provider base64Provider, RequestEncode.Provider bitmapProvider) {
        this.context = context;
        this.service = service;
        this.base64Provider = base64Provider;
        this.bitmapProvider = bitmapProvider;
    }

    private static Base64Image init(Context context, ExecutorService service, RequestDecode.Provider base64Provider, RequestEncode.Provider bitmapProvider) {
        if (singleton == null) {
            synchronized (Base64Image.class) {
                if (singleton == null) {
                    singleton = new Base64Image(context, service, base64Provider, bitmapProvider);
                }
            }
        }
        return singleton;
    }

    public static Base64Image getInstance() {
        return singleton;
    }

    public RequestDecode decode(String path) {
        if (path.trim().length() == 0) {
            throw new IllegalArgumentException("Path must not be empty.");
        }
        return new RequestDecode(this, path, base64Provider);
    }

    public RequestEncode encode(String path) {
        if (path.trim().length() == 0) {
            throw new IllegalArgumentException("Path must not be empty.");
        }
        return new RequestEncode(this, path, bitmapProvider);
    }

    public static final class Builder {
        private Context context;
        private ExecutorService service;
        private RequestDecode.Provider base64Provider;
        private RequestEncode.Provider bitmapProvider;

        public Builder() {
        }

        public Base64Image build() {

            if (service == null) {

                //service = new Base64ExecutorService();

                float scaleFactor = 0.5f;
                int cpuCount = Runtime.getRuntime().availableProcessors();
                int maxThreads = (int) (cpuCount * scaleFactor);
                maxThreads = (maxThreads > 0 ? maxThreads : 1);

                service = new ThreadPoolExecutor(
                        maxThreads, // core thread pool size
                        maxThreads, // maximum thread pool size
                        60, TimeUnit.SECONDS,
                        new LinkedBlockingDeque<Runnable>(),
                        new ThreadPoolExecutor.CallerRunsPolicy());
            }

            return Base64Image.init(context, service, base64Provider, bitmapProvider);
        }

        public Builder setContext(Context context) {
            this.context = context.getApplicationContext();
            return this;
        }

        public Builder setBase64Provider(RequestDecode.Provider provider) {
            this.base64Provider = provider;
            return this;
        }

        public Builder setBitmapProvider(RequestEncode.Provider provider) {
            this.bitmapProvider = provider;
            return this;
        }
    }
}

File: src/main/java/com/pixelcarrot/base64image/Base64Image.kt
package com.pixelcarrot.base64image

import android.graphics.Bitmap
import java.util.concurrent.LinkedBlockingDeque
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

object Base64Image {

    private const val SCALE = 0.5F
    private const val KEEP_ALIVE_TIME = 60L

    private val cpuCount by lazy { Runtime.getRuntime().availableProcessors() }
    private val thread by lazy { (cpuCount * SCALE).toInt() }
    private val maxThreads by lazy { if (thread > 0) thread else 1 }

    private val service by lazy {
        ThreadPoolExecutor(
            maxThreads, // core thread pool size
            maxThreads, // maximum thread pool size
            KEEP_ALIVE_TIME,
            TimeUnit.SECONDS,
            LinkedBlockingDeque(),
            ThreadPoolExecutor.CallerRunsPolicy()
        )
    }

    @JvmStatic
    fun encode(bitmap: Bitmap?, callback: (base64: String?) -> Unit) {
        val job = bitmap?.let {
            Encoder(bitmap, object : SimpleListener<String?>() {
                override fun onCompleted(result: String?) {
                    callback(result)
                }
            })
        }
        job?.let { service.submit(job) } ?: run { callback(null) }
    }

    @JvmStatic
    fun decode(base64: String?, callback: (bitmap: Bitmap?) -> Unit) {
        val job = base64?.let {
            Decoder(base64, object : SimpleListener<Bitmap?>() {
                override fun onCompleted(result: Bitmap?) {
                    callback(result)
                }
            })
        }
        job?.let { service.submit(job) } ?: run { callback(null) }
    }
}